File: Java_ChatGPT/Chain_of_responsbility/Round2/Code.java
package Round2;

// DeliveryRequest class that holds details of the package delivery
class DeliveryRequest {
    private String location;
    private boolean paymentProcessed;
    private boolean delivered;

    public DeliveryRequest(String location) {
        this.location = location;
        this.paymentProcessed = false;
        this.delivered = false;
    }

    public String getLocation() {
        return location;
    }

    public boolean isPaymentProcessed() {
        return paymentProcessed;
    }

    public void setPaymentProcessed(boolean paymentProcessed) {
        this.paymentProcessed = paymentProcessed;
    }

    public boolean isDelivered() {
        return delivered;
    }

    public void setDelivered(boolean delivered) {
        this.delivered = delivered;
    }
}

// Abstract handler in the chain
abstract class Handler {
    protected Handler nextHandler;

    public void setNextHandler(Handler nextHandler) {
        this.nextHandler = nextHandler;
    }

    public abstract void handleRequest(DeliveryRequest request);
}

// LocationHandler verifies if the location is valid for delivery
class LocationHandler extends Handler {
    private static final String VALID_LOCATION = "ValidLocation";

    @Override
    public void handleRequest(DeliveryRequest request) {
        if (VALID_LOCATION.equals(request.getLocation())) {
            System.out.println("Location is valid.");
            if (nextHandler != null) {
                nextHandler.handleRequest(request);
            }
        } else {
            System.out.println("Invalid location. Cannot proceed.");
        }
    }
}

// PaymentHandler processes payment
class PaymentHandler extends Handler {
    @Override
    public void handleRequest(DeliveryRequest request) {
        if (!request.isPaymentProcessed()) {
            System.out.println("Processing payment...");
            request.setPaymentProcessed(true);
            if (nextHandler != null) {
                nextHandler.handleRequest(request);
            }
        } else {
            System.out.println("Payment already processed.");
        }
    }
}

// DeliveryHandler manages the actual delivery of the package
class DeliveryHandler extends Handler {
    @Override
    public void handleRequest(DeliveryRequest request) {
        if (request.isPaymentProcessed()) {
            System.out.println("Package delivered.");
            request.setDelivered(true);
        } else {
            System.out.println("Cannot deliver package. Payment not processed.");
        }
    }
}

// Main class to set up and test the Chain of Responsibility
public class Code {
    public static void main(String[] args) {
        // Create handlers
        Handler locationHandler = new LocationHandler();
        Handler paymentHandler = new PaymentHandler();
        Handler deliveryHandler = new DeliveryHandler();

        // Set up the chain
        locationHandler.setNextHandler(paymentHandler);
        paymentHandler.setNextHandler(deliveryHandler);

        // Create a valid delivery request
        DeliveryRequest validRequest = new DeliveryRequest("ValidLocation");
        locationHandler.handleRequest(validRequest);

        // Create an invalid delivery request
        DeliveryRequest invalidRequest = new DeliveryRequest("InvalidLocation");
        locationHandler.handleRequest(invalidRequest);
    }
}

File: Python_Gemini_Flash/Chain_of_responsbility/Round1/Code.py
class Handler:
  """Abstract base class for handlers in the chain"""
  def __init__(self, next_handler=None):
    self.next_handler = next_handler

  def handle(self, request):
    """Processes the request or passes it to the next handler"""
    processed = self._handle(request)
    if not processed and self.next_handler:
      processed = self.next_handler.handle(request)
    return processed

  def _handle(self, request):
    """Concrete implementation of request processing for each handler"""
    raise NotImplementedError("Subclass must implement _handle")

class LocationHandler(Handler):
  """Verifies if the delivery can be made to the specified location"""
  def _handle(self, request):
    if request.location == "Restricted Area":
      print(f"Delivery cannot be made to {request.location}")
      return False
    else:
      print(f"Delivery location {request.location} is valid.")
      return True

class PaymentHandler(Handler):
  """Handles payment processing for the delivery"""
  def _handle(self, request):
    if not request.payment_processed:
      print("Processing payment...")
      # Simulate payment processing
      request.payment_processed = True
      print("Payment successful.")
      return True
    else:
      print("Payment already processed.")
      return True

class DeliveryHandler(Handler):
  """Manages the actual delivery of the package"""
  def _handle(self, request):
    if request.location_verified and request.payment_processed:
      print(f"Delivering package to {request.location}")
      return True
    else:
      print("Delivery cannot proceed yet.")
      return False

# Example usage
def deliver_package(request):
  """Initializes the chain of responsibility and starts the delivery process"""
  first_handler = LocationHandler(
      PaymentHandler(DeliveryHandler())
  )
  return first_handler.handle(request)

# Sample request data
request = {
  "location": "Customer Address",
  "payment_processed": False
}

deliver_package(request)
